comp/otelcol/otlp/config.go

```diff
@@ -62,17 +64,63 @@ func readConfigSection(cfg config.Reader, section string) *confmap.Conf {
 	for _, key := range cfg.AllKeysLowercased() {
 		if strings.HasPrefix(key, prefix) && cfg.IsSet(key) {
 			mapKey := strings.ReplaceAll(key[len(prefix):], ".", confmap.KeyDelimiter)
+
+			if v, ok := typeMap[mapKey]; ok {
+				if v == "bool" {
+					stringMap[mapKey] = cfg.GetBool(key)
+					continue
+				}
+				if v == "int" {
+					stringMap[mapKey] = cfg.GetInt(key)
+					continue
+				}
+			}
+
 			// deep copy since `cfg.Get` returns a reference
 			stringMap[mapKey] = deepcopy.Copy(cfg.Get(key))
 		}
 	}
 	return confmap.NewFromStringMap(stringMap)
 }
 
+// typeTagMap takes a struct and returns a map with mapstructure tag as key and field type as value.
+func typeTagMap(input interface{}, lead string, result map[string]string) map[string]string {
+	v := reflect.ValueOf(input)
+	t := v.Type()
+
+	// Check if the input is a struct
+	if t.Kind() != reflect.Struct {
+		return result
+	}
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+
+		// Get the mapstructure tag, skip if no tag is set
+		tag := field.Tag.Get("mapstructure")
+		if tag == "" || tag == "-" {
+			continue
+		}
+		if tag == "omitempty" || tag == ",squash" {
+			tag = ""
+		}
+		if field.Type.Kind() == reflect.Struct {
+			typeTagMap(v.Field(i).Interface(), lead+tag, result)
```

Comment:
`lead` typically would be named `prefix` as per other code that works with the config. `lead` also is not being used in this function except for recursive calls. Did you mean to use it [here](https://github.com/DataDog/datadog-agent/pull/30666/files#diff-154d9b9bc61ef1bb8c3caeb7364b3e7f7c9f80f00147edeaa25f8f035164853bR114) as well?